sdks/python/apache_beam/transforms/util.py

```diff
@@ -785,7 +788,6 @@ def expand(self, pcoll):
 
   @experimental()
   @typehints.with_input_types(Tuple[K, V])
-  @typehints.with_output_types(Tuple[K, Iterable[V]])
```

Comment:
Any reason why this isn't:
```python
  @typehints.with_output_types(Tuple[ShardedKeyType[K], Iterable[V]])
```
?

Reply:
Because I got an error saying:
```
arg = ShardedKey[int], msg = 'Tuple[t0, t1, ...]: each t must be a type.', is_argument = True

    def _type_check(arg, msg, is_argument=True):
        """Check that the argument is a type, and return it (internal helper).
    
        As a special case, accept None and return type(None) instead. Also wrap strings
        into ForwardRef instances. Consider several corner cases, for example plain
        special forms like Union are not valid, while Union[int, str] is OK, etc.
        The msg argument is a human-readable error message, e.g::
    
            "Union[arg, ...]: arg should be a type."
    
        We append the repr() of the actual value (truncated to 100 chars).
        """
        invalid_generic_forms = (Generic, Protocol)
        if is_argument:
            invalid_generic_forms = invalid_generic_forms + (ClassVar, Final)
    
        if arg is None:
            return type(None)
        if isinstance(arg, str):
            return ForwardRef(arg)
        if (isinstance(arg, _GenericAlias) and
                arg.__origin__ in invalid_generic_forms):
            raise TypeError(f"{arg} is not valid as type argument")
        if (isinstance(arg, _SpecialForm) and arg not in (Any, NoReturn) or
                arg in (Generic, Protocol)):
            raise TypeError(f"Plain {arg} is not valid as type argument")
        if isinstance(arg, (type, TypeVar, ForwardRef)):
            return arg
        if not callable(arg):
>           raise TypeError(f"{msg} Got {arg!r:.100}.")
E           TypeError: Tuple[t0, t1, ...]: each t must be a type. Got ShardedKey[int].

../../../../../../.pyenv/versions/3.8.5/lib/python3.8/typing.py:149: TypeError

```
Any insights?

Reply:
Right, Python's `typing` types don't recognize Beam's own internal types. 

1.
Probably not a solution:
ShardedKeyTypeConstraint could inherit from `typing.Generic[typing.TypeVar('K')]`, but that produces 2 more problems. Neither are hard but add complexity, plus none of the other Beam typehints do this.
- Accessing the type argument that Generic stores on different versions of Python. (`__args__` or `__parameters__` I believe)
- Pickling. Python's typing types have had issues with pickling.

2.
Probably the right solution but breaks consistency with the other decorators:

```py
@typehints.with_output_types(typehints.Tuple[ShardedKeyType[K], typehints.Iterable[V]])
```

Beam already converts to internal type so this is a cosmetic change.

Reply:
Ah that explains. To avoid the decorator inconsistency I still keep the `infer_output_type` method but instead use `typehint.Tuple` as opposed to `TupleConstraint` directly. Is that acceptable?

Reply:
I don't believe you need infer_output_type if you have the with_output_types decorator.

Also, I tested this a bit myself and it ended up looking like this:
```py
  @typehints.with_input_types(typehints.Tuple[K_i, V_i])
  @typehints.with_output_types(typehints.Tuple[ShardedKeyType[K_i], typehints.Iterable[V_i]])
  class WithShardedKey(PTransform):
```

where:
```py
K_i = typehints.TypeVariable('K_i')
V_i = typehints.TypeVariable('V_i')
```

Reply:
Sorry I misunderstood and thought having something like `typehints.Tuple` in the decoration was not preferred. Pushed a commit to use decoration instead.

Reply:
It's not preferred, but I'd rather have the output type documented there than not at all.